radix-engine/src/system/system_db_reader.rs

```diff
@@ -140,6 +285,254 @@ impl<'a, S: SubstateDatabase> SystemDatabaseReader<'a, S> {
+        let type_info = self.fetch_substate::<SpreadPrefixKeyMapper, TypeInfoSubstate>(
+            node_id,
+            TYPE_INFO_FIELD_PARTITION,
+            &TypeInfoField::TypeInfo.into(),
+        )?;
+
+        let kv_store_info = match type_info {
+            TypeInfoSubstate::KeyValueStore(kv_store_info) => kv_store_info,
+            _ => return None,
+        };
+
+        Some(KVStoreTypeTarget {
+            kv_store_type: kv_store_info.generic_substitutions,
+            meta: *node_id,
+        })
+    }
+
+    pub fn get_blueprint_type_target(
+        &self,
+        node_id: &NodeId,
+        module_id: ObjectModuleId,
+    ) -> Option<BlueprintTypeTarget> {
+        let type_info = self.fetch_substate::<SpreadPrefixKeyMapper, TypeInfoSubstate>(
+            node_id,
+            TYPE_INFO_FIELD_PARTITION,
+            &TypeInfoField::TypeInfo.into(),
+        )?;
+
+        let object_info = match type_info {
+            TypeInfoSubstate::Object(object_info) => object_info,
+            _ => return None,
+        };
+
+        if let Some(_version) = object_info.module_versions.get(&module_id) {
+            match module_id {
+                ObjectModuleId::Main => {
+                    let target = BlueprintTypeTarget {
+                        blueprint_info: object_info.blueprint_info,
+                        meta: SchemaValidationMeta::ExistingObject {
+                            additional_schemas: *node_id,
+                        },
+                    };
+                    Some(target)
+                }
+                _ => {
+                    let blueprint_id = module_id.static_blueprint().unwrap();
+                    let target = BlueprintTypeTarget {
+                        blueprint_info: BlueprintInfo {
+                            blueprint_id,
+                            outer_obj_info: OuterObjectInfo::None,
+                            features: Default::default(),
+                            generic_substitutions: Default::default(),
+                        },
+                        meta: SchemaValidationMeta::ExistingObject {
+                            additional_schemas: *node_id,
+                        },
+                    };
+                    Some(target)
+                }
+            }
+        } else {
+            None
+        }
+    }
+
+    pub fn get_kv_store_payload_schema(
+        &self,
+        target: &KVStoreTypeTarget,
+        key_or_value: KeyOrValue,
+    ) -> Option<ResolvedPayloadSchema> {
+        let (substs, allow_ownership, allow_non_global_refs) = match key_or_value {
+            KeyOrValue::Key => (
+                &target.kv_store_type.key_generic_substitutions,
+                false,
+                false,
+            ),
+            KeyOrValue::Value => (
+                &target.kv_store_type.value_generic_substitutions,
+                target.kv_store_type.allow_ownership,
+                false,
```

Comment:
Should be depending on if the blueprint is transient?

Reply:
Not sure what you mean here?

Reply:
The third element of the tuple `allow_non_global_ref` should depend on if the owning object is transient or not, rather than the fact this is a key value store?

Reply:
Ah, for the case of key value stores this is hardcoded (especially since no current transient object has key value stores). We could formalize this by saying "non global refs are only allowed in the schema defined parts of a transient object"